chain/epoch_manager/src/shard_assignment.rs

```diff
@@ -11,21 +11,26 @@ use std::cmp;
 /// best when the number of chunk producers is greater than
 /// `num_shards * min_validators_per_shard`.
 pub fn assign_shards<T: HasStake + Eq + Clone>(
-    chunk_producers: Vec<T>,
+    mut chunk_producers: Vec<T>,
     num_shards: NumShards,
     min_validators_per_shard: usize,
 ) -> Result<Vec<Vec<T>>, NotEnoughValidators> {
-    // Initially, sort by number of validators, then total stake
-    // (i.e. favour filling under-occupied shards first).
-    let mut shard_validator_heap: MinHeap<(usize, Balance, ShardId)> =
-        (0..num_shards).map(|s| (0, 0, s)).collect();
-
     let num_chunk_producers = chunk_producers.len();
     if num_chunk_producers < min_validators_per_shard {
         // Each shard must have `min_validators_per_shard` distinct validators,
         // however there are not that many, so we cannot proceed.
         return Err(NotEnoughValidators);
     }
+
+    // We’re using greedy algorithm so make sure chunk producers are sorted from
+    // the one with biggest stake.
+    chunk_producers.sort_by(|a, b| b.get_stake().cmp(&a.get_stake()));
```

Comment:
I think the chunk producers here is already sorted, because of how `select_validators` works. It should rank validator stake from high to low and select by order.

I think it's probably better to add in the specs of `assign_shards` that it assumes `chunk_producers` are sorted by stake and adds a check in the function that checks the chunk producer stake is in decreasing order.

Reply:
Checking is actually more code than simply sorting it turns out (`is_sorted_by` is a nightly feature) and `sort_by` is O(n) if sequence is already sorted.  With that I think that we should either sort or document and I’d prefer the former since that will guard against any potential in the future where someone forgets that the function expects producers to be ordered. Thoughts?

Reply:
In my personal opinion, I would still prefer the check it's the same cost as sort and it is not that complicated, and it does not add unnecessary code that may cause confusion for people who are reading the code. Plus, I don't think the `assign_shards` function will be used by any where else in which it is not taking input from `select_validators` (the check actually serves as a guard for implementation errors in `select_validators`), expect in tests.

Reply:
PTAL.